src/ncurses_ui.cc

```diff
@@ -661,12 +661,51 @@ Optional<Key> NCursesUI::get_next_key()
         const int new_c = wgetch(m_window);
         if (new_c == '[') // potential CSI
         {
-            const Codepoint csi_val = wgetch(m_window);
-            switch (csi_val)
+            const Codepoint c1 = wgetch(m_window);
+            switch (c1)
             {
                 case 'I': return {Key::FocusIn};
                 case 'O': return {Key::FocusOut};
-                default: break; // nothing
+                case '1':
+                {
+                  const Codepoint c2 = wgetch(m_window);
+                  if ( c2 != ';' )
+                  {
+                      ungetch(c2); ungetch(c1); break;
+                  }
+
+                  const Codepoint c3 = wgetch(m_window);
+                  Key (*f)(Key) = NULL;
+                  switch (c3)
+                  {
+                      case '2': f = shift; break;
+                      case '3': f = alt; break;
+                      case '4': f = shift_alt; break;
+                      case '5': f = ctrl; break;
+                      case '6': f = shift_ctrl; break;
+                      case '7': f = alt_ctrl; break;
+                      case '8': f = shift_alt_ctrl; break;
+                  }
+                  if ( f == NULL )
+                  {
+                      ungetch(c3); ungetch(c2); ungetch(c1); break;
+                  }
```

Comment:
Check if the functor (from above) has a valid target, instead of checking for the `NULL` value (in C++, you should rather use `std::nullptr`, but it won't matter with a functor).

Reply:
I've switched to `nullptr`. I didn't embed this check into the switch, because we don't want the `break` on the inner one, we want it on the outer one.

Reply:
nullptr is a built in, so no need for std:: in front. but you are correct, we use nullptr in Kakoune codebase.